### ISSUE TITLE
2: Deployment (Practical Deep Learning for Coders)

### ISSUE THREAD

**Issue by JeremyLoh** [JeremyLoh/ml#6]
https://course.fast.ai/Lessons/lesson2.html
- Design own machine learning project, creating your own dataset, training a model with your data, and deploying an application on the web
- We will use a particular deployment target called [Hugging Face Space](https://huggingface.co/spaces) with [Gradio](https://gradio.app/), and will also see how to use JavaScript to implement an interface in the browser.

# Kaggle notebook used for own practice - https://www.kaggle.com/code/jeremyloh42/chapter-2-deep-learning-for-coders

# Tasks
- [x] Read Chapter 2 of Book (Deep Learning for Coders with Fastai and PyTorch) - **Start Page 138 / 1028**
    - Current Page: 184
- [x] Read Chapter 2 Notebook https://github.com/fastai/fastbook/blob/master/02_production.ipynb
- [x] 50:00 / 1:16:41 Lesson 2: Practical Deep Learning for Coders 2022 https://www.youtube.com/watch?v=F4tvM4Vb3A0

https://aiquizzes.com/

https://forums.fast.ai/

independent variable vs dependent variable
- The independent variable is the cause. Its value is independent of other variables in your study. The dependent variable is the effect. Its value depends on changes in the independent variable

What is the difference between item_tfms and batch_tfms?
- item_tfms is item transform applied on an individual item basis. This is done on the CPU. batch_tfms is batch transform applied on batches of data. This is done in GPU.
- https://www.educative.io/answers/what-is-data-block-api-in-fastai

# Book Page 159
For dataloaders, `item_tfms`
```python
bears = DataBlock(
    blocks=(ImageBlock, CategoryBlock),
    get_items=get_image_files,
    splitter=RandomSplitter(valid_pct=0.2, seed=42),
    get_y=parent_label,
    item_tfms=Resize(128))
```
By default, Resize crops the images to fit a square shape of the size requested, using the full width or height. This can result in losing some important details.

Alternatively, you can ask fastai to pad the images with zeros (black), or
squish/stretch them

All of these approaches seem somewhat wasteful or problematic. If we squish or stretch the images, they end up as unrealistic shapes, leading to a model that learns that things look different from how they actually are, which we would expect to result in lower accuracy.

If we crop the images, we remove some of the features that allow us to perform recognition. For instance, if we were trying to recognize breeds
of dog or cat, we might end up cropping out a key part of the body or the face necessary to distinguish between similar breeds.

If we pad the images, we have a whole lot of empty space, which is just wasted computation for our model and results in a lower effective resolution for the part of the image we actually use.

![Image](https://user-images.githubusercontent.com/38811217/263475493-37220fc0-323e-4202-8518-dfc1db7295f0.png)

![Image](https://user-images.githubusercontent.com/38811217/263475476-aa99a2cf-2d85-4ef3-b60b-9e89eed42bde.png)

![Image](https://user-images.githubusercontent.com/38811217/263475701-42f04f09-b646-4b10-b735-04aac9ac4579.png)

When you call export, fastai will save a file called export.pkl:
learn.export()

When we use a model for getting predictions, instead of training, we call it
**inference**. To create our inference learner from the exported file, we use
load_learner

![Image](https://user-images.githubusercontent.com/38811217/263524356-9ed1bc9f-2604-410c-afce-b1a6f8b7bdd5.png)

### Potential issues (page 177/1028 of pdf book)
**Out of domain data** - there may be data that our model sees in production that is very different from what it saw during training
- We have to be careful about our approach to rolling out the technology

**Domain shift** - where the type of data that our model sees changes over time. e.g. insurance company may use a deep learning model as part of its pricing and risk algorithm. Over time, the types of customers the company attracts and the types of risk might change so much that the original training data is no longer relevant

A larger problem is that we can never fully understand all the possible behaviors of a neural network, because they have **far too many parameters**

Deployment process
1. Manual process
    - Run model in parallel
    - Humans check all predictions 
2. Limited scope deployment
    - Careful human supervision
    - Time or geography limited 
3. Gradual expansion
    - Good reporting systems needed
    - Consider what could go wrong 

## Video Chapter 2 notes

7:10 - fastai library has GUI for data cleaning `ImageClassifierCleaner(learn)` that allows you to choose a category and the training versus validation set and view the highest-loss images in order. It has menus to allow images to be selected for removal or relabelling.
![Image](https://user-images.githubusercontent.com/38811217/263425457-02b48cb7-b9fa-4042-8cf5-966f91277033.png)

12:45 - **Before you clean the dataset, you train the model**

19:00 - Confusion matrix, only is meaningful when label has categories. It shows what category errors we are making. Shows which ones are hard to identify.
```python
interp = ClassificationInterpretation.from_learner(learn)
interp.plot_confusion_matrix()
```

20:40
```python
interp = ClassificationInterpretation.from_learner(learn)
interp.plot_top_losses(5, nrows=1, figsize=(17,4))
```
Tells places where the loss is the highest. Loss is the measurement of how good the model is, we check this loss after we run through each epoch of training
Loss is bad if we predict badly and we are very confident about our prediction
If prediction is correct, but it is not confident (probability), it is also bad

Bad loss
- Bad prediction (Wrong) and confident (probability)
- Right prediction but not confident

22:10
A model can help you find data issues more quickly and easily. So we train a simple model quickly first and use it to help us with data cleaning

We can use fastai GUI for data cleaning `ImageClassifierCleaner` that allows us to choose a category and the training versus validation set, and view the highest-loss images in order. It has menus that allow images to be selected for removal or relabelling

```python
cleaner = ImageClassifierCleaner(learn) # it runs our model to get greatest loss data
cleaner

# choose images

for idx in cleaner.delete():
    cleaner.fns[idx].unlink()

for idx, category in cleaner.change():
    shutil.move(str(cleaner.fns[idx]), path/category)
```

ImageClassifierCleaner doesn't actually delete or change labels for you. It just returns the indices of items to change.

**Before starting data cleaning, build a model to find out what things are difficult to recognize in your data and find data problems**

25:30
On computer, normal RAM doesn't get filled up because if you use too much RAM, your computer will start swapping (swap to hard disk to use for later)
GPU when run out of RAM, no other options, can't swap

29:20
Put model into Hugging Face spaces server and write a UI for it

39:00
```python
learn.export("model.pkl") # create a file with info to run the model, pickle file
```

41:15
How to do predictions on saved model ("model.pkl" file)?
- Any external functions used in labelling, need to also created. Learner saved the name of function used, but doesn't have the function definition.
![Image](https://user-images.githubusercontent.com/38811217/263438428-de2c9268-afdf-4956-9d61-7fdae13af402.png)

Jupyter has metrics, e.g. `%time learn.predict(im)` shows how long something took

48:30
We can export Juypter cells into a script using `#|export` for the cells we want to export, then we can run
```python
#|default_exp app
# ^ this tells what is the default python file name to export as
from nbdev.export import notebook2script
notebook2script("app.ipynb") #name of this notebook to parse
```

54:30
Getting Python, Jupyter notebooks and fastai running on local machine
https://github.com/fastai/fastsetup
- Clone the repo and then run `./setup-conda.sh` which install python
- Don't use system Python (python install on the system, to do other tools for computer). So we have own development version
- The script installs mambaforge, so we can do `mamba install fastai` (after flosing and opening the terminal
- **Refer to https://docs.fast.ai/#Installing**, replace `conda` with `mamba`, its much faster

```shell
mamba install -c fastchan fastai

mamba install -c fastchan nbdev
```

```shell
jupyter notebook --no-browser
```
